libscript/include/pivot/script/Parser.hxx

```diff
@@ -0,0 +1,61 @@
+#pragma once
+
+#include "pivot/script/DataStructures.hxx"
+
+#include <stack>
+
+namespace pivot::ecs::script::parser
+{
+
+//	Main function, will return the abstract syntax tree of the file (root Node)
+Node ast_from_file(const std::string &file, bool isContent = false, bool verbose = true);
+
+// Lexer function, will return a list of Tokens contained in the file
+std::vector<Token> tokens_from_file(const std::string &file, bool isContent = false, bool verbose = false);
+
+// private functions
+Node consumeComponent(
+    std::vector<Token> &tokens);    // Consume a component token and all following to build a component declaration node
+void consumeComponentToken(std::vector<Token> &tokens, Node &result, TokenType expectedType, NodeType fillType,
+                           Token &lastToken);    // consume one token
+Node consumeSystem(
+    std::vector<Token> &tokens);    // Consume a system token and all following to build a system declaration node
+void consumeSystemDescriptionToken(std::vector<Token> &tokens, Node &result, TokenType expectedType, NodeType fillType,
+                                   Token &lastToken);    // consume one token for declaration
+void consumeSystemStatement(std::vector<Token> &tokens, Node &result,
+                            Token &lastToken);    // consume an entire statement and append it to children node
+void consumeSystemBlock(std::vector<Token> &tokens, Node &result,
+                        Token &lastToken);    // consume entire block and append it to children node
+void consumeSystemVariable(std::vector<Token> &tokens, Node &result,
+                           Token &lastToken);    // consume a variable and append it to children node
+void consumeSystemExpression(std::vector<Token> &tokens, Node &result,
+                             Token &lastToken);    // consume an expression and append it to children node
+void consumeSystemFuncParams(std::vector<Token> &tokens, Node &result,
+                             Token &lastToken);    // consume function parameters and append them to children node
+
+void expectSystemToken(
+    std::vector<Token> &tokens, TokenType expectedType, Token &lastToken,
+    bool consume);    // check that token exists, and is of correct type (and potentially consume it from tokens)
+void expectSystemTokenValue(
+    std::vector<Token> &tokens, const std::string &expectedValue, Token &lastToken,
+    bool consume);    // check that token exists, and is of correct value (and potentially consume it from tokens)
+bool isDeclarationOver(const std::vector<Token> &tokens);    // Check for end of system/component declaration
```

Comment:
Could you wrap all of that into a function with method ?

You almost always carry around a `std::vector<Token>`, that could be a member variable, and everything else could be method.